test/test_database.py

```diff
@@ -138,29 +138,133 @@ def db(self, dependency_provider):
+
         commit.assert_called()
         rollback.assert_not_called()
         close.assert_called()
 
+    @patch.object(Session, 'rollback')
+    @patch.object(Session, 'commit')
+    @patch.object(Session, 'close')
+    def test_rolls_back(self, close, commit, rollback, db):
+
+        with pytest.raises(Exception):
+            with db.get_session():
+                raise Exception('Yo!')
+
+        commit.assert_not_called()
+        rollback.assert_called()
+        close.assert_not_called()
+
     @patch.object(Session, 'rollback')
     @patch.object(Session, 'commit')
     @patch.object(Session, 'close')
     def test_rolls_back_and_closes(self, close, commit, rollback, db):
 
         with pytest.raises(Exception):
-            with db.get_session() as session:
-                assert isinstance(session, Session)
+            with db.get_session(close_on_exit=True):
                 raise Exception('Yo!')
 
         commit.assert_not_called()
         rollback.assert_called()
         close.assert_called()
 
+    @patch.object(Session, 'rollback')
+    @patch.object(Session, 'commit')
+    @patch.object(Session, 'close')
+    def test_rolls_back_on_commit_error(
+        self, close, commit, rollback, db
+    ):
+
+        commit.side_effect = Exception('Yo!')
+
+        with pytest.raises(Exception):
+            with db.get_session():
+                pass
+
+        rollback.assert_called()
+        close.assert_not_called()
+
+    @patch.object(Session, 'rollback')
+    @patch.object(Session, 'commit')
+    @patch.object(Session, 'close')
+    def test_rolls_back_and_closes_on_commit_error(
+        self, close, commit, rollback, db
+    ):
+
+        commit.side_effect = Exception('Yo!')
+
+        with pytest.raises(Exception):
+            with db.get_session(close_on_exit=True):
+                pass
+
+        rollback.assert_called()
+        close.assert_called()
+
+    @patch.object(Session, 'rollback')
+    @patch.object(Session, 'commit')
+    @patch.object(Session, 'close')
+    def test_no_close_on_exit(
+        self, close, commit, rollback, db
+    ):
+
+        with db.get_session() as session_one:
+            assert isinstance(session_one, Session)
+
+        commit.assert_called()
+        rollback.assert_not_called()
+        close.assert_not_called()
+
+        with db.get_session(close_on_exit=True) as session_two:
+            assert isinstance(session_two, Session)
+
+        assert commit.call_count == 2
+        rollback.assert_not_called()
+        close.assert_called()
+
+        assert db._context_sessions == [session_one, session_two]
+
+        assert close.call_count == 1
+        db.close()
+        assert close.call_count == 3
+
+    def test_worker_teardown(self, dependency_provider):
+        dependency_provider.setup()
+
+        worker_ctx = Mock(spec=WorkerContext)
+        db = dependency_provider.get_dependency(worker_ctx)
+
+        with db.get_session(close_on_exit=False) as session_one:
```

Comment:
`close_on_exit` defaults to `False` now. Were these cases intended to be different?

Reply:
true, it should be true.)

I should test final closing - either on exit or by worker teardown

I changed both adding the argument to explicitly say.